README
# Rail Map Generator V2

Demo page: https://wongchito.github.io/RailMapGeneratorV2/

**Notice:**

- Supported browser: Google Chrome (> 74.0) **ONLY**
- Supported feature:
    - Change theme colour from selected mass transport system

Previous version: [GenRailMap](https://github.com/wongchito/GenRailMap) (to be deprecated)

CHANGELOG.md
# Changelog

## 3.8.35 (05 Jul 2021)

### Resolved Issues

- Change MTR Tuen Ma Line Phase 1 from colour palette to Tuen Ma Line
- Update Tuen Ma Line template, marks the opening of Sung Wong Toi and To Kwa Wan stations

## 3.8.26 (07 Jul 2021)

### New Features

-   New runin terminal style of Shanghai Metro 2020 (#83).
-   Parallel destination name of Shanghai Metro 2020 (#83).
-   Missing maglev in the template of Shanghai Metro line 2 (#85).
-   Update the template of Shanghai Metro line 1.

## 3.8.25 (23 Mar 2021)

### New Features

-   Shanghai Metro's 2020 new style (#68).

## 3.8.20 (8 Nov 2021)

### New Features

-   New templates are available: Shanghai Metro Line 2, Shanghai Metro Line 5 and Shanghai Metro Line 16.

## 3.8.18 (25 Oct 2021)

### New Features

-   Shanghai Metro Style supports **Maglev** icon.
-   Shanghai Metro Style supports **Only for Public Transportation Card** hint.

## 3.8 (20 Jun 2020)

### New Features

-   Added express and direct services for Shanghai Metro style.
-   New templates are available: Shanghai Metro Line 1 and Shanghai Metro Line 16.

## 3.7.2 (15 May 2020)

### New Features

-   Updated TypeScript to 3.9.2.
-   Added Prettier to local dependencies.

### Resolved Issues

-   Fixed typos.

## 3.7.1 (22 Apr 2020)

### Resolved Issues

-   Fixed the issue where interchange boxes mis-align in Shanghai Metro style.
-   Fixed [#42](https://github.com/wongchito/RailMapGenerator/issues/42).

## 3.7 (14 Apr 2020)

### New Features

-   Stations' names can now be placed below the line when they are not staggered in MTR style.
-   Added selections of recently used colour-name combinations in colour editing dialogue. ([#25](https://github.com/wongchito/RailMapGenerator/issues/25))
-   Supported platform number in Shanghai Metro style.

### Resolved Issues

-   Simplified paths (strip) with `<marker>`.
-   Refined sizes, aspect ratios and alignments of some elements in Shanghai Metro style.

## 3.6.7 (13 Apr 2020)

### New Features

-   New palettes: Kansai Area and Osaka
-   Redesign layout of colour editing dialogue.

### Resolved Issues

-   Fixed the issue where font size in exported PNG is incorrect.
-   Fixed the issue where font are not correctly rendered in exported PNG.

## 3.6.6 (11 Apr 2020)

### New Features

-   Supported panel type of Guangzhou Metro Line 2/8 (Otis).
-   ~~Supported progressive web app (PWA) (beta).~~

### Resolved Issues

-   Fixed alignment issue in destination sign of Shanghai Metro style.
-   Reordered canvas's background and border.
-   Improved algorithms of loading rendered fonts in MTR style.

## 3.6.5 (6 Apr 2020)

### New Features

-   Added Terms and Conditions in the preview dialogue.

### Resolved Issues

-   Improved decoration line under stations' names in Shanghai Metro style.

## 3.6.4 (5 Apr 2020)

### New Features

-   Rotations of stations' names in Shanghai Metro style are now determined by train direction.
-   New theme colours for Madrid.

### New Features

## 3.6.3 (5 Apr 2020)

### New Features

-   New theme colours for Oslo, Stockholm and Tyne and Wear.
-   Added decoration line under names of stations whose are interchange stations in Shanghai Metro style. (#29)

### Resolved Issues

-   Fixed the issues where OSI text is not removed when OSI has been removed in Shanghai Metro style. (#29)

## 3.6.2 (2 Mar 2020)

### New Features

-   New design of 'Stations' panel.
-   New theme colours for Barcelona.

### Resolved Issues

-   Fixed an error where invalid attributes are passed to elements when secondary name is empty in Guangzhou Metro style.

## 3.6.1 (31 Mar 2020)

### Resolved Issues

-   Fixed an error where exports from Firefox are empty. (#26)
-   Improved compatibility for Safari and Firefox.

## 3.6 (30 Mar 2020)

### New Features

-   New interface of station name editing dialogue.
-   Secondary station name is supported in Guangzhou Metro style.
-   New theme colours for Nanjing, Paris, Qingdao, Tianjin.

## 3.5.6 (29 Mar 2020)

### Resolved Issues

-   Fixed an error that occurs when reversing stations more than once.
-   Caught errors in exporting process.
-   Improved robustness.

## 3.5.5 (28 Mar 2020)

### New Features

-   New design of export and preview dialog, which allows you to customise the styles (including transparent background, border and scaling) of the exported images. (#2)
-   New theme colours for Hangzhou.

### Resolved Issues

-   Fixed a typo in panel type's description.

## 3.5.4 (27 Mar 2020)

### New Features

-   You may now stop staggering stations' names in MTR style and place them above the line.
-   New template available (MTR Disneyland Resort Line).

### Resolved Issues

-   Fixed an error in rail map where the background of current station has incorrect size in MTR style.
-   Fixed an error in destination sign where possible terminals are incorrect.

## 3.5.3 (27 Mar 2020)

### Resolved Issues

-   Completed type of `RMGParam` and get rid of index signatures.
-   Simplified parameter update process.

## 3.5.2 (27 Mar 2020)

### New Features

-   Added types of information panels for Line 4, Line 5 and Guangfo Line Phase 1 in Guangzhou Metro style.
-   Updated information panel type selector.
-   Deployed on [GitLab Pages](https://chitowong.gitlab.io/RailMapGenerator) as a backup plan.

### Resolved Issues

-   Fixed an backward compatibility issue in running-in board of Shanghai Metro style.

## 3.5.1 (24 Mar 2020)

### New Features

-   You may now switch between styles without refreshing the whole web page.
-   Improved layout for tablet/desktop users.

### Resolved Issues

-   Fixed an issue in design panel where sliders of note's position don't respond in Guangzhou Metro style.

## 3.5 (22 Mar 2020)

### New Features

-   Update environment to `create-react-app` set-up. This includes the following changes:
    -   New directories of assets, and
    -   Strict mode in TypeScript.
-   Added error boundary to SVG components.

### Resolved Issues


Station.js
'use strict';

// var colours;
// $.ajax({
//     url: 'colours.json', 
//     success: function (data) {
//         colours = data;
//     }, 
//     async: false
// });

class Station {
    _x; _y; _id; _state; 
    _namePos;
    // constructor (id, par, child, names) {
    //     this._id = id;
    //     this._parents = par;
    //     this._children = child;
    //     [this._nameZH, this._nameEN] = names;
    // }

    constructor (id, data) {
        this._id = id;
        this._parents = data.parents;
        this._children = data.children;
        [this._nameZH, this._nameEN] = data.name;
    }

    set x(val) {this._x = val;}
    // get x() {return this._x;}

    set y(val) {this._y = val;}
    // get y() {return this._y;}

    set _state(val) {this._state = val;}

    set namePos(pos) {
        // 0: upper, 1: lower
        this._namePos = pos;
    }

    get nameClass() {
        switch (this._state) {
            case -1:
                return 'Pass';
            case 0:
                return 'Current';
            default:
                return 'Future';
        }
    }

    get nameHTML() {
        var dy = (this._namePos == 1) ? 12 + 11.843775 : -12 - 21.921875;
        switch (this._state) {
            case -1:
                var nameClass = 'Pass';
                break;
            case 0:
                var nameClass = 'Current';
                break;
            default:
                var nameClass = 'Future';
        }
        return `<g text-anchor="middle" class="Name ${nameClass}" transform="translate(${this._x},${this._y + dy})">
                <text class="StnNameZH"> ${this._nameZH} </text> 
                <text dy="15" class="StnNameEN"> ${this._nameEN} </text> 
                </g>`;
    }

    get iconHTML() {
        var iconType = (this._state == -1) ? 'stn_hk_pass' : 'stn_hk';
        return `<use xlink:href="#${iconType}" transform="translate(${this._x},${this._y})"/>`;
    }

    get ungrpHTML() {
        return this.iconHTML + this.nameHTML;
    }

    get html() {
        return `<g id="${this._id}"> ${this.ungrpHTML} </g>`;
    }
}

class Int2Station extends Station {
    #intCity; #intLine; #intColour;
    #intNameZH; #intNameEN;
    // constructor (id, par, child, names, int2) {
    //     super(id, par, child, names);
    //     [this.#intCity, this.#intLine, this.#intColour, this.#intNameZH, this.#intNameEN] = int2;
    //     // this.#intColour = colours[this.#intCity]['line'][this.#intLine]['colour'];
    // }

    constructor (id, data) {
        super(id, data);
        [this.#intCity, this.#intLine, this.#intColour, this.#intNameZH, this.#intNameEN] = data.transfer[1];
    }

    get _dy() {return 0;}

    get intTickHTML() {
        var tickDirection = (this._namePos == 1) ? 'up' : 'down';
        var tickColour = (this._state == -1) ? '#aaa' : this.#intColour;
        return `<use xlink:href="#intline_${tickDirection}" class="LineDiagram HK Change" stroke="${tickColour}" x="${this._x}" y="${this._y + this._dy}"/>`;
    }

    get _nameClass() {
        return (this._state == -1) ? 'Pass' : 'Future';
    }

    get intNameHTML() {
        var [nameHTML, nameZHLn, nameENLn] = joinIntName([this.#intNameZH, this.#intNameEN], 15, 7);
        var dy = (this._namePos == 0) ? 25 + 5.953125 : -25 + 5.953125 - 18.65625 - 13*(nameZHLn-1) - 7*(nameENLn-1);
        dy += this._dy;
        // var nameClass = (this._state == -1) ? 'Pass' : 'Future';
        return `<g text-anchor="middle" class="Name ${this._nameClass}" transform="translate(${this._x},${this._y + dy})">
                ${nameHTML}
                </g>`;
    }

    get ungrpHTML() {
        return this.intTickHTML + this.iconHTML + this.nameHTML + this.intNameHTML;
    }
}

class Int3Station extends Station {
    #intCity = []; 
    #intLine = []; 
    #intColour = [];
    #intNameZH = []; 
    #intNameEN = [];
    _int3Type;

    constructor (id, data) {
        super(id, data);

        data.transfer.slice(1).forEach(intInfo => {
            this.#intCity.push(intInfo[0]);
            this.#intLine.push(intInfo[1]);
            this.#intColour.push(intInfo[2]);
            // this.#intColour.push(colours[intInfo[0]]['line'][intInfo[1]]['colour']);

            this.#intNameZH.push(intInfo[3]);
            this.#intNameEN.push(intInfo[4]);
        });

        this._int3Type = data.change_type.substring(5);
    }

    get iconHTML() {
        var iconType = (this._state == -1) ? 'int3_hk_pass' : 'int3_hk';
        var iconRotation = (this._namePos != 1) ? 0 : 180;
        return `<use xlink:href="#${iconType}" transform="translate(${this._x},${this._y})rotate(${iconRotation})"/>`;
    }

    get _tickRotation() {return 0;}
    get _dy() {return 0;}

    get intTickHTML() {
        // var tickRotation = (this._int3Type == 'r') ? -90 : 90;
        var str = '';
        [0, 1].forEach(i => {
            var tickColour = (this._state == -1) ? '#aaa' : this.#intColour[i];
            var dy = (this._namePos != 1) ? 18*(i+1) : -18*(i+1);
            dy += this._dy;
            str += `<use xlink:href="#intline_down" class="LineDiagram HK Change" stroke="${tickColour}"
            transform="translate(${this._x},${this._y + dy})rotate(${this._tickRotation})"/>`
        });
        return str;
    }

    get _txtAnchor() {return 'middle';}
    get _intNameDX() {return 0;}

    get intNameHTML() {
        var str = '';
        var nameClass = (this._state == -1) ? 'Pass' : 'Future';

        [0, 1].forEach(i => {
            var [nameHTML, nameZHLn, nameENLn] = joinIntName([this.#intNameZH[i], this.#intNameEN[i]], 15, 7)
            // var dy = (this._namePos == 0) ? 25 + 5.953125 : -25 + 5.953125 - 19.65625 - 13*(nameZHLn-1) - 7*(nameENLn-1);
            var dy = ( (this._namePos == 0) ? 18*(i+1) : -18*(i+1) ) + 5.953125 - (19.65625 + 13*(nameZHLn-1) + 7*(nameENLn-1))/2;
            dy += this._dy;
            str += `<g text-anchor="${this._txtAnchor}" class="Name ${nameClass}" transform="translate(${this._x + this._intNameDX},${this._y + dy})">
                    ${nameHTML}
                    </g>`
        });
        return str;
    }

    get ungrpHTML() {
        return this.intTickHTML + this.iconHTML + this.nameHTML + this.intNameHTML;
    }
}

class Int3LStation extends Int3Station {
    get _tickRotation() {return 90;}
    get _txtAnchor() {return 'end';}
    get _intNameDX() {return -24;}
}

class Int3RStation extends Int3Station {
    get _tickRotation() {return -90;}
    get _txtAnchor() {return 'start';}
    get _intNameDX() {return 24;}
}

class OSI11Station extends Int2Station {
    #osiNameZH; #osiNameEN; 
    // #intCity; #intLine; #intColour; 
    // #intNameZH; #intNameEN;
    #osiType; #osiDirection;
    constructor (id, data) {
        // data.int2 = data.osi11;
        super(id, data);

        [this.#osiNameZH, this.#osiNameEN] = data.transfer[0];

        this.#osiType = data.change_type.substring(6,7); // u(npaid) or p(aid);
        this.#osiDirection = data.change_type.substring(7); // l or r;
    }

    get iconHTML() {
        var iconType = `osi11${this.#osiType}_hk${(this._state == -1) ? '_pass' : ''}`;
        var iconRotation = (this._namePos != 1) ? 0 : 180;
        return `<use xlink:href="#${iconType}" transform="translate(${this._x},${this._y})rotate(${iconRotation})"/>`;
    }

    get _dy() {return (this._namePos == 0) ? 26 : -26;}
    get _txtAnchor() {return 'middle';}
    get _osiNameDX() {return 0;}

    get osiNameHTML() {
        var dy = this._dy + 8.34375 - 25.03125/2;
        return `<g text-anchor="${this._txtAnchor}" transform="translate(${this._x+this._osiNameDX},${this._y+dy})" class="Name ${this._nameClass}">
                <text class="OSINameZH"> ${this.#osiNameZH} </text>
                <text x="0" dy="12" class="OSINameEN"> ${this.#osiNameEN} </text>
                </g>`;
    }

    get ungrpHTML() {
        return this.intTickHTML + this.iconHTML + this.nameHTML + this.intNameHTML + this.osiNameHTML;
    }
}

class OSI11LStation extends OSI11Station {
    // OSI name on left
    get _txtAnchor() {return 'end';}
    get _osiNameDX() {return -13;}
}

class OSI11RStation extends OSI11Station {
    // OSI name on right
    get _txtAnchor() {return 'start';}
    get _osiNameDX() {return 13;}
}

class OSI12Station extends Int3Station {
    #osiNameZH; #osiNameEN; 
    #osiType; #osiDirection;
    constructor (id, data) {
        // data.int3 = data.osi12;
        super(id, data);

        [this.#osiNameZH, this.#osiNameEN] = data.transfer[0];

        this.#osiType = data.change_type.substring(6,7); // u(npaid) or p(aid);
        this.#osiDirection = data.change_type.substring(7); // l or r;
    }

    get iconHTML() {
        var iconType = `osi12${this.#osiType}_hk${(this._state == -1) ? '_pass' : ''}`;
        var iconRotation = (this._namePos != 1) ? 0 : 180;
        return `<use xlink:href="#${iconType}" transform="translate(${this._x},${this._y})rotate(${iconRotation})"/>`;
    }

    get _dy() {return (this._namePos == 0) ? (26-18) : -8;}

    get osiNameHTML() {
        var dy = (this._namePos == 0) ? (26+18+10) + 8.34375  : -(26+18+10) + 8.34375 - 25.03125;
        var nameClass = (this._state == -1) ? 'Pass' : 'Future';
        return `<g text-anchor="middle" transform="translate(${this._x},${this._y+dy})" class="Name ${nameClass}">
                <text class="OSINameZH"> ${this.#osiNameZH} </text>
                <text x="0" dy="12" class="OSINameEN"> ${this.#osiNameEN} </text>
                </g>`;
    }

    get ungrpHTML() {
        return this.intTickHTML + this.iconHTML + this.nameHTML + this.intNameHTML + this.osiNameHTML;
    }
}

class OSI12LStation extends OSI12Station {
    get _tickRotation() {return 90;}
    get _txtAnchor() {return 'end';}
    get _intNameDX() {return -24;}
}

class OSI12RStation extends OSI12Station {
    get _tickRotation() {return -90;}
    get _txtAnchor() {return 'start';}
    get _intNameDX() {return 24;}
}